### PR TITLE
#9 downloadable data

### PR DESCRIPTION
**Pull request by ratulkai93**
This PR allows the user to choose from the provided file formats and download datasets in that format

image

TXT
image

JSON
image

CSV
image

XLS
image

### CODE REVIEW

**Review by Mehrdadmaskull — APPROVED** [review]
Nice and clean code, LGTM

**Comment by Mehrdadmaskull** [comment]
Just make sure the duplication the Sonar is saying is fixed

File: Client/src/Components/DatasetUpload/CSVdatasetDownloadButton.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+        for (var i in datasets.material) {
+            materials += datasets.material[i].composition + " \'" + datasets.material[i].details + "\', "
+        }
+        materials = materials.slice(0, -2)
+        return materials
+    }
+    const authorArray = (): any => {
+        var authors = ""
+        for (var i in datasets.reference.authors) {
+            if (datasets.reference.authors[i].middleName)
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/CSVdatasetDownloadButton.tsx]: Good check

File: Client/src/Components/DatasetUpload/CSVdatasetDownloadButton.tsx [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+            "#Category, " + categoryName() + "\n" +
+            "#Subcategory, " + subcategoryName() + "\n" +
+            "#\n" +
+            "#Publications/Source\n" +
+            "#Authors, " + authorArray() + "\n" +
+            "#Title, " + datasets.reference.title + "\n" +
+            "#Type, " + datasets.reference.type + "\n" +
+            "#Publisher, " + datasets.reference.publisher + "\n"
+            if (datasets.reference.volume) { output += "#Volume, " + datasets.reference.volume + "\n" }
+            if (datasets.reference.pages) { output += "#Pages, " + datasets.reference.pages + "\n" }
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/CSVdatasetDownloadButton.tsx]: Nice null handling

File: Client/src/Components/DatasetUpload/JSONdatasetDownloadButton.tsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+                )
+            }
+
+        }
+        return name
+    }
+    const dataContentArray = (): any => {
+        var name = []
+        for (var i in datasets.data.contents) {
+            if (datasets.data.dataPointComments)
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/JSONdatasetDownloadButton.tsx:75]: Nice null handling 2

File: Client/src/Components/DatasetUpload/JSONdatasetDownloadButton.tsx [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+        "Data type": datasets.data_type,
+        "Category": categoryName(),
+        "Subcategory ": subcategoryName(),
+        "reference": {
+            "Authors ": [authorArray()],
+            "Title ": datasets.reference.title,
+            "Type": datasets.reference.type,
+            "Publisher": datasets.reference.publisher,
+            "Volume, ": datasets.reference.volume,
+            "Pages, ": datasets.reference.pages,
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/JSONdatasetDownloadButton.tsx:105]: Would be nice to have null handling here, but not feasible given time constraints

File: Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+            materials += datasets.material[i].composition + " \'" + datasets.material[i].details + "\', "
+        }
+        materials = materials.slice(0, -2)
+        return materials
+    }
+
+    const authorArray = (): any => {
+        var authors = ""
+        for (var i in datasets.reference.authors) {
+            if (datasets.reference.authors[i].middleName)
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx]: Return of null checks

File: Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+        return authors
+    }
+    const dataContentArray = (): any => {
+        var dataPoint = ""
+        for (var i in datasets.data.contents) {
+            dataPoint += datasets.data.contents[i].point + ", "
+            if (datasets.data.dataPointComments)
+                dataPoint += datasets.data.dataPointComments[i] + "\n"
+            else
+                dataPoint = dataPoint.slice(0, -2) + "\n"
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx]: Didn't mention this earlier, but a good way to get rid of trailing commas

File: Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+            "#Subcategory: " + subcategoryName() + "\n" +
+            "#\n" +
+            "#Publications/Source:\n" +
+            "#Authors: " + authorArray() + "\n" +
+            "#Title: " + datasets.reference.title + "\n" +
+            "#Type: " + datasets.reference.type + "\n" +
+            "#Publisher: " + datasets.reference.publisher + "\n"
+        if (datasets.reference.volume) { output += "#Volume: " + datasets.reference.volume + "\n" }
+        if (datasets.reference.pages) { output += "#Pages: " + datasets.reference.pages + "\n" }
+        if (datasets.reference.doi) { output += "#DOI: " + datasets.reference.doi + "\n" }
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/TXTdatasetDownloadButton.tsx]: Null check returns again

File: Client/src/Components/DatasetUpload/XLSdatasetDownloadButton.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+                cat.subcategories.map(subcategory => {
+                    if (subcategory.id == datasets.subcategory) {
+                        scgName.push(subcategory.name)
+                    }
+                })
+            }
+        })
+        return scgName
+    }
+    const compileDataArray = (): any => {
```
Comment (silent-po2) [inline Client/src/Components/DatasetUpload/XLSdatasetDownloadButton.tsx:83]: Slightly messy, but it works. Without time constraints something cleaner could be looked into, but alas

**Review by silent-po2 — APPROVED** [review]
Good stuff, ready to go once Sonar is sorted out

**Pull request merged**